ops/testing.py

```diff
@@ -2434,66 +2513,63 @@ def push(
+            os.chmod(file_path, permissions)
         except FileNotFoundError as e:
             raise pebble.PathError(
                 'not-found', f'parent directory not found: {e.args[0]}')
-        except NonAbsolutePathError as e:
-            raise pebble.PathError(
-                'generic-file-error',
-                f'paths must be absolute, got {e.args[0]!r}'
-            )
+        except NotADirectoryError:
+            raise pebble.PathError('generic-file-error',
+                                   f'open {path}.~: not a directory')
 
     def list_files(self, path: str, *, pattern: Optional[str] = None,
                    itself: bool = False) -> List[pebble.FileInfo]:
         self._check_connection()
-        try:
-            files = [self._fs.get_path(path)]
-        except FileNotFoundError:
-            # conform with the real pebble api
+        self._check_absolute_path(path)
+        file_path = self._root / path[1:]
+        if not file_path.exists():
             raise pebble.APIError(
                 body={}, code=404, status='Not Found',
                 message=f"stat {path}: no such file or directory")
-
+        files = [file_path]
         if not itself:
             try:
-                files = self._fs.list_dir(path)
+                files = [file_path / file for file in os.listdir(file_path)]
             except NotADirectoryError:
                 pass
 
         if pattern is not None:
             files = [file for file in files if fnmatch.fnmatch(file.name, pattern)]
 
-        type_mappings = {
-            _File: pebble.FileType.FILE,
-            _Directory: pebble.FileType.DIRECTORY,
-        }
-
-        def get_pebble_file_type(file: '_FileOrDir') -> pebble.FileType:
-            pebble_type = type_mappings.get(type(file))
-            if not pebble_type:
-                raise ValueError(
-                    f'unable to convert file {file} (type not one of {type_mappings})')
-            return pebble_type
-
-        return [
-            pebble.FileInfo(
-                path=str(file.path),
-                name=file.name,
-                type=get_pebble_file_type(file),
-                size=file.size if isinstance(file, _File) else None,
-                permissions=file.kwargs.get('permissions') or 0,
-                last_modified=file.last_modified,
-                user_id=file.kwargs.get('user_id'),
-                user=file.kwargs.get('user'),
-                group_id=file.kwargs.get('group_id'),
-                group=file.kwargs.get('group'),
-            )
+        file_infos = [
+            Container._build_fileinfo(file)
             for file in files
         ]
+        for file_info in file_infos:
+            rel_path = os.path.relpath(file_info.path, start=self._root)
+            rel_path = '/' if rel_path == '.' else '/' + rel_path
+            file_info.path = rel_path
+            file_info.name = "/" if rel_path == "/" else os.path.basename(rel_path)
```

Comment:
Why isn't `file_info.name` correct already?

Reply:
if it is only for the root, why set it always, vs just:
```
if rel_path == "/":
  // override the root name, since it should be considered the local root, without any filename
  file_info.name = "/"
  ```
 
It seems clearer than always setting it to the value it should already have. (and avoids computing basename() for every file in every directory if you are iterating over them)

Reply:
John's suggestion looks clearer to me too.